### ISSUE TITLE
MSBuild CodeGenerator chokes on generic method in grain interface (simple repo)

### ISSUE THREAD

**Issue by oising** [dotnet/orleans#6719]
The following code will repo the bug on v3.2.2 (latest as I write this.) I'm using dotnet 5.0p8 (but fairly certain it also fails also on 3.1) 

```csharp
using System;
using System.Threading.Tasks;
using Orleans;

namespace OrleansCodeGenBug
{
    public interface IThing
    {
        Task<int> DoSomething<TGrain>() where TGrain : IOtherGrain;
    }

    public interface ISimpleGrain : IGrainWithStringKey, IThing
    {
        Task<string> Greet();
    }

    public class SimpleGrain : Grain, ISimpleGrain
    {
        public Task<string> Greet()
        {
            return Task.FromResult("Hello, world.");
        }

        // IThing -- this also fails to codegen if I implement explicitly
        public Task<int> DoSomething<TGrain>() where TGrain: IOtherGrain
        {
            var other = GrainFactory.GetGrain<TGrain>("foo");

            return Task.FromResult(0);
        }
    }

    public interface IOtherGrain : IGrainWithStringKey {  }
}
```

Error:

```text
1>You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
1>Orleans.CodeGenerator - command-line = SourceToSource C:\p\OrleansCodeGenBug\obj\Debug\netstandard2.1\OrleansCodeGenBug.orleans.g.args.txt
1>-- Code Generation FAILED --
1>
1>Exc level 0: System.Collections.Generic.KeyNotFoundException: The given key 'OrleansCodeGenBug.IThing.DoSomething<TGrain>()' was not present in the dictionary.
1>   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
1>   at Orleans.CodeGenerator.Generators.GrainMethodInvokerGenerator.GenerateInvokeForMethod(WellKnownTypes wellKnownTypes, ExpressionSyntax castGrain, IMethodSymbol method, ExpressionSyntax arguments, Dictionary`2 genericInvokerFields) in D:\build\agent\_work\24\s\src\Orleans.CodeGenerator\Generators\GrainMethodInvokerGenerator.cs:line 249
1>   at Orleans.CodeGenerator.Generators.GrainMethodInvokerGenerator.<>c__DisplayClass8_0.<GenerateInvokeMethod>b__1(IMethodSymbol methodType) in D:\build\agent\_work\24\s\src\Orleans.CodeGenerator\Generators\GrainMethodInvokerGenerator.cs:line 204
1>   at Orleans.CodeGenerator.Generators.GrainInterfaceCommon.GenerateGrainInterfaceAndMethodSwitch(WellKnownTypes types, INamedTypeSymbol grainType, ExpressionSyntax methodIdArgument, Func`2 generateMethodHandler, Func`3 composeInterfaceBlock) in D:\build\agent\_work\24\s\src\Orleans.CodeGenerator\Generators\GrainInterfaceCommon.cs:line 62
1>   at Orleans.CodeGenerator.Generators.GrainMethodInvokerGenerator.GenerateInvokeMethod(WellKnownTypes wellKnownTypes, INamedTypeSymbol grainType, IMethodSymbol invokeMethod, Dictionary`2 genericInvokerFields) in D:\build\agent\_work\24\s\src\Orleans.CodeGenerator\Generators\GrainMethodInvokerGenerator.cs:line 200
1>   at Orleans.CodeGenerator.Generators.GrainMethodInvokerGenerator.GenerateInvokeMethod(WellKnownTypes wellKnownTypes, INamedTypeSymbol grainType, Dictionary`2 genericInvokerFields) in D:\build\agent\_work\24\s\src\Orleans.CodeGenerator\Generators\GrainMethodInvokerGenerator.cs:line 111
1>   at Orleans.CodeGenerator.Generators.GrainMethodInvokerGenerator.GenerateClass(GrainInterfaceDescription description) in D:\build\agent\_work\24\s\src\Orleans.CodeGenerator\Generators\GrainMethodInvokerGenerator.cs:line 65
1>   at Orleans.CodeGenerator.CodeGenerator.GenerateSyntax(AggregatedModel model) in D:\build\agent\_work\24\s\src\Orleans.CodeGenerator\CodeGenerator.cs:line 145
1>   at Orleans.CodeGenerator.CodeGenerator.GenerateCode(CancellationToken cancellationToken) in D:\build\agent\_work\24\s\src\Orleans.CodeGenerator\CodeGenerator.cs:line 59
1>   at Orleans.CodeGenerator.MSBuild.CodeGeneratorCommand.Execute(CancellationToken cancellationToken) in D:\build\agent\_work\24\s\src\Orleans.CodeGenerator.MSBuild\CodeGeneratorCommand.cs:line 143
1>   at Microsoft.Orleans.CodeGenerator.MSBuild.Program.SourceToSource(String[] args) in D:\build\agent\_work\24\s\src\Orleans.CodeGenerator.MSBuild\Program.cs:line 138
1>   at Microsoft.Orleans.CodeGenerator.MSBuild.Program.Main(String[] args) in D:\build\agent\_work\24\s\src\Orleans.CodeGenerator.MSBuild\Program.cs:line 26
1>C:\Users\OisinGrehan\.nuget\packages\microsoft.orleans.codegenerator.msbuild\3.2.2\build\Microsoft.Orleans.CodeGenerator.MSBuild.targets(125,5): error MSB3073: The command ""dotnet" "C:\Users\OisinGrehan\.nuget\packages\microsoft.orleans.codegenerator.msbuild\3.2.2\build\..\tasks\netcoreapp3.1\Orleans.CodeGenerator.MSBuild.dll" SourceToSource "C:\p\OrleansCodeGenBug\obj\Debug\netstandard2.1\OrleansCodeGenBug.orleans.g.args.txt"" exited with code 3.
1>Done building project "OrleansCodeGenBug.csproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 1 up-to-date, 0 skipped ==========
```

**Comment by oising** [comment]
Huh, so it's 1am and after playing around trying to figure what's going on, I shifted IGrainWithStringKey over to IThing, and the codegen elicited a message about needing to include Microsoft.Orleans.Core to handle generic methods. I added it, and sure enough this new problem went away. So then I switched back to the above code, and it was also fixed :) So I guess hiding generic methods in base interfaces is not triggering your error method. I'll leave this open to track the fix, and to leave the explanation for others to find.

**Comment by ReubenBond** [comment]
Good find, and thanks for reporting and experimenting. We can keep this open until we have a fix

**Comment by annemartijn0** [comment]
Using Microsoft.Orleans.Core instead of Microsoft.Orleans.Core.Abstractions also resolved this issue for me.

**Comment by ReubenBond** [comment]
Fixed in 4.0.0-preview1
